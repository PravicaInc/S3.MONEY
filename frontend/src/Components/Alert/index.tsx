import { FC, ReactNode } from 'react';
import { twMerge } from 'tailwind-merge';

import AlertIcon from '@/../public/images/alert_icon.svg?jsx';
import BackgroundModalDecorativeIcon from '@/../public/images/background_modal_decorative_left.svg?jsx';

import { Button } from '@/Components/Form/Button';
import { Modal, ModalProps } from '@/Components/Modal';

export interface AlertProps extends ModalProps {
  onOkClick: () => void;
  okButtonContent?: ReactNode;
  header?: ReactNode;
  description?: ReactNode;
}

export const Alert: FC<AlertProps> = ({
  onClose,
  onOkClick,
  header,
  description,
  okButtonContent = 'OK',
  className,
  ...props
}) => (
  <Modal
    onClose={onClose}
    className={twMerge('relative p-6 w-[400px]', className)}
    {...props}
  >
    <div className="absolute top-0 left-0 z-[-1]">
      <BackgroundModalDecorativeIcon />

      <div className="absolute top-6 left-6 bg-mistyRose w-12 h-12 flex items-center justify-center rounded-full">
        <AlertIcon />
      </div>
    </div>
    <p className="text-primary text-lg font-semibold mt-16">
      {header}
    </p>
    <p className="mt-1 text-secondary">
      {description}
    </p>
    <Button
      className="mt-8 w-full h-11"
      onClick={onOkClick}
    >
      {okButtonContent}
    </Button>
  </Modal>
);
