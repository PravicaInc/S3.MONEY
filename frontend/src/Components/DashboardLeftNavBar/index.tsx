'use client';

import { FC, HTMLAttributes, useMemo } from 'react';
import Link from 'next/link';
import { usePathname, useSearchParams } from 'next/navigation';
import { twMerge } from 'tailwind-merge';

import LogoIcon from '@/../public/images/logo.svg?jsx';
import OperationsLinkIcon from '@/../public/images/operations_link_icon.svg?jsx';

import { PAGES_URLS } from '@/utils/const';

export const DashboardLeftNavBar: FC<HTMLAttributes<HTMLDivElement>> = ({ className, ...props }) => {
  const searchParams = useSearchParams();
  const pathname = usePathname();

  const links = useMemo(
    () => [
      {
        groupName: 'Main Menu',
        groupLinks: [
          {
            href: {
              pathname: PAGES_URLS.dashboardOperations,
              query: Object.fromEntries(searchParams.entries()),
            },
            icon: <OperationsLinkIcon />,
            text: 'Operations',
            isActive: pathname === PAGES_URLS.dashboardOperations,
          },
        ],
      },
    ],
    [searchParams, pathname]
  );

  return (
    <div
      className={twMerge(
        'h-screen w-[270px] border-r border-borderPrimary bg-white',
        className
      )}
      {...props}
    >
      <Link
        href={PAGES_URLS.home}
        className="ml-6 my-5 block"
      >
        <LogoIcon />
      </Link>
      {
        links.map(({ groupName, groupLinks }) => (
          <div key={groupName} className="pt-5 px-4">
            <p className="text-xs font-semibold text-hitGrey pl-3 pb-3">
              {groupName}
            </p>
            <div className="flex flex-col gap-1">
              {groupLinks.map(({ text, href, icon, isActive }) => (
                <Link
                  key={text}
                  href={href}
                  className={twMerge(
                    'flex items-center gap-[10px] py-2 px-3 border rounded-md border-transparent text-secondary',
                    isActive && 'border-borderPrimary bg-[#F8F9FB] text-primary'
                  )}
                >
                  {icon}
                  {text}
                </Link>
              ))}
            </div>
          </div>
        ))
      }
    </div>
  );
};
