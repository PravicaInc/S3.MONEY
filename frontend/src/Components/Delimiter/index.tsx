import { FC, HTMLAttributes } from 'react';
import { twMerge } from 'tailwind-merge';

export const Delimiter: FC<HTMLAttributes<HTMLDivElement>> = ({ className, ...props }) => (
  <div
    className={twMerge(
      'bg-borderPrimary h-[1px]',
      className
    )}
    {...props}
  />
);
