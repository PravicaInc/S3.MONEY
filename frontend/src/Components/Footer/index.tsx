import { FC, HTMLAttributes } from 'react';
import { twMerge } from 'tailwind-merge';

export const Footer: FC<HTMLAttributes<HTMLDivElement>> = ({ className, ...props }) => (
  <div
    className={twMerge('font-medium flex items-center justify-between text-mistBlue', className)}

    {...props}
  >
    <p>
      Powered By
      {' '}
      <a href="https://pravica.io/" target="_blank" rel="noreferrer">
        Pravica
      </a>
    </p>
    <p>
      Copyright
      {' '}
      {(new Date()).getFullYear()}
    </p>
  </div>
);
