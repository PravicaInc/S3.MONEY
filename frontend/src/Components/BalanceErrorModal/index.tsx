import { FC, useCallback, useState } from 'react';
import { useSuiClientContext } from '@mysten/dapp-kit';

import AlertIcon from '@/../public/images/alert_icon.svg?jsx';
import BackgroundModalDecorativeIcon from '@/../public/images/background_modal_decorative_left.svg?jsx';

import { Button, BUTTON_VIEWS } from '@/Components/Form/Button';
import { Modal, ModalProps } from '@/Components/Modal';

import { useRequestSuiTokens } from '@/hooks/useRequestSuiTokens';

export interface BalanceErrorModalProps extends ModalProps {}

export const BalanceErrorModal: FC<BalanceErrorModalProps> = ({
  onClose,
  ...props
}) => {
  const suiClientContext = useSuiClientContext();
  const requestSuiTokens = useRequestSuiTokens();

  const [requestSuiTokensSuccess, setRequestSuiTokensSuccess] = useState<boolean>(false);
  const [requestSuiTokensError, setRequestSuiTokensError] = useState<boolean>(false);

  const requestToken = useCallback(
    async () => {
      const faucetRequestStatus = await requestSuiTokens.mutateAsync();

      setRequestSuiTokensSuccess(faucetRequestStatus === 'SUCCEEDED');
      setRequestSuiTokensError(faucetRequestStatus === 'DISCARDED');
    },
    [requestSuiTokens]
  );

  return (
    <Modal
      onClose={onClose}
      className="relative p-6 w-[480px]"
      {...props}
    >
      <div className="absolute top-0 left-0 z-[-1]">
        <BackgroundModalDecorativeIcon />

        <div className="absolute top-6 left-6 bg-mistyRose w-12 h-12 flex items-center justify-center rounded-full">
          <AlertIcon />
        </div>
      </div>
      <p className="text-primary text-lg font-semibold mt-16">
        Your balance is too low
      </p>
      <p className="mt-1 text-secondary">
        Unfortunately, your balance is too small to create a new StableCoin.
        {
          requestSuiTokensError && (
            <>
              <br />
              <span className="text-error">
                Something went wrong, and we were unable to complete your request. Please try again.
              </span>
            </>
          )
        }
      </p>
      <div className="grid grid-cols-4 items-center justify-center gap-4 mt-8">
        <Button
          className="h-[56px] col-span-1"
          view={BUTTON_VIEWS.secondary}
          onClick={onClose}
        >
          Close
        </Button>
        {
          ['testnet', 'devnet'].includes(suiClientContext.network) && (
            <Button
              className="h-[56px] col-span-3"
              onClick={requestToken}
              disabled={requestSuiTokens.isPending || requestSuiTokens.isSuccess}
              isLoading={requestSuiTokens.isPending}
            >
              {
                requestSuiTokensSuccess
                  ? (
                    <>
                      {
                        suiClientContext.network === 'testnet'
                          ? 1
                          : 10
                      }
                      {' '}
                      SUI received
                    </>
                  )
                  : (
                    <>
                      Request
                      <span className="capitalize px-1">
                        {suiClientContext.network}
                      </span>
                      SUI Tokens
                    </>
                  )
              }
            </Button>
          )
        }
      </div>
    </Modal>
  );
};
