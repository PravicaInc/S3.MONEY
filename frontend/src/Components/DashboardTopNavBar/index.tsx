'use client';

import { FC, HTMLAttributes, useMemo } from 'react';
import { useAutoConnectWallet, useCurrentAccount } from '@mysten/dapp-kit';
import { twMerge } from 'tailwind-merge';

import { Loader } from '@/Components/Loader';
import { LogoutButton } from '@/Components/LogoutButton';
import { SelectStableCoinDropdown } from '@/Components/SelectStableCoinDropdown';

import { useShortAccountAddress } from '@/hooks/useShortAccountAddress';

export const DashboardTopNavBar: FC<HTMLAttributes<HTMLDivElement>> = ({ className, ...props }) => {
  const autoConnectionStatus = useAutoConnectWallet();
  const account = useCurrentAccount();
  const shortAccountAddress = useShortAccountAddress();

  const isLoading = useMemo(
    () => autoConnectionStatus === 'idle',
    [autoConnectionStatus]
  );
  const isRedirecting = useMemo(
    () => autoConnectionStatus === 'attempted' && !account?.address,
    [autoConnectionStatus, account?.address]
  );

  return (
    <div
      className={twMerge(
        'flex items-center justify-between p-6 h-24 border-b border-borderPrimary bg-white',
        className
      )}
      {...props}
    >
      <SelectStableCoinDropdown />
      {
        (isLoading || isRedirecting) && (
          <Loader className="h-8" />
        )
      }
      {
        shortAccountAddress && (
          <LogoutButton />
        )
      }
    </div>
  );
};
