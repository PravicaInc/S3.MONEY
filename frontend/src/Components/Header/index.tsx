'use client';

import { FC, HTMLAttributes, useCallback, useState } from 'react';
import NextImage from 'next/image';
import { usePathname, useRouter } from 'next/navigation';
import { twMerge } from 'tailwind-merge';

import LogoIcon from '@/../public/images/logo.svg?jsx';
import LogoutIcon from '@/../public/images/logout.svg?jsx';

import { Loader } from '@/Components/Loader';
import { LogoutModal } from '@/Components/LogoutModal';

import { PAGES_URLS } from '@/utils/const';

import { useWallet } from '@/hooks/useWallet';

export const Header: FC<HTMLAttributes<HTMLDivElement>> = ({ className, ...props }) => {
  const wallet = useWallet();
  const router = useRouter();
  const pathname = usePathname();

  const [showLogoutModal, setShowLogoutModal] = useState<boolean>(false);

  const disconnect = useCallback(
    () => {
      wallet.disconnect();
      router.replace(`${PAGES_URLS.signIn}?next=${encodeURIComponent(pathname)}`);
    },
    [wallet, router, pathname]
  );

  return (
    <div
      className={twMerge(
        'min-h-20 px-8 border-b border-primaryBorder bg-white',
        className
      )}
      {...props}
    >
      <div
        data-testid="header"
        className="flex items-center justify-between max-w-screen-2xl mx-auto h-full"
      >
        <LogoIcon />
        {
          (wallet.connecting || wallet.disconnected) && (
            <Loader className="h-8" />
          )
        }
        {
          wallet.connected && wallet.shortWalletAddress && (
            <button
              className="
                border border-primaryBorder rounded-[10px]
                h-10 px-4 flex items-center justify-between gap-2
                text-secondary font-medium
                transition
                hover:bg-actionPrimary hover:bg-opacity-10 hover:border-actionSecondary
              "
              onClick={() => {
                setShowLogoutModal(true);
              }}
            >
              {
                wallet.adapter?.icon && (
                  <NextImage
                    className="rounded-full"
                    src={wallet.adapter?.icon}
                    alt="Wallet Icon"
                    width={24}
                    height={24}
                  />
                )
              }
              {wallet.shortWalletAddress}
              <LogoutIcon />
            </button>
          )
        }
      </div>
      <LogoutModal
        visible={showLogoutModal}
        onClose={() => setShowLogoutModal(false)}
        inProcess={wallet.connecting || wallet.disconnected}
        onProceed={disconnect}
      />
    </div>
  );
};
