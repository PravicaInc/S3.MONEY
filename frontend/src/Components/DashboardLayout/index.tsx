import { FC, ReactNode } from 'react';
import { twMerge } from 'tailwind-merge';

import { DashboardLeftNavBar } from '@/Components/DashboardLeftNavBar';
import { DashboardTopNavBar } from '@/Components/DashboardTopNavBar';
import { Footer } from '@/Components/Footer';

export interface DashboardLayoutProps {
  children?: ReactNode;
  className?: string;
  contentClassName?: string;
}

export const DashboardLayout: FC<DashboardLayoutProps> = ({ children, className, contentClassName }) => (
  <div className={twMerge('flex min-h-screen bg-pageBgPrimary', className)}>
    <DashboardLeftNavBar className="shrink-0" />
    <div className={twMerge('w-screen h-screen flex flex-col', contentClassName)}>
      <DashboardTopNavBar className="shrink-0" />
      <div className="overflow-auto h-full flex flex-col justify-between">
        <div className="h-full">
          {children}
        </div>
        <Footer
          className="pb-7 pl-9 pr-[85px]"
        />
      </div>
    </div>
  </div>
);

