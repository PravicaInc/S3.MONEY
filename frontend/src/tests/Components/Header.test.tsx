import { useAutoConnectWallet, useCurrentAccount } from '@mysten/dapp-kit';
import { act, cleanup, getQueriesForElement, RenderResult } from '@testing-library/react';

import '@testing-library/jest-dom';

import { Header } from '@/Components/Header';

import { createMockAccount, TEST_SHORT_WALLET_ACCOUNT_ADDRESS } from '@/tests/utils/create_mock_wallet_account';
import { renderWithProviders } from '@/tests/utils/render_with_providers';

const renderHeaderWithProviders = () => renderWithProviders(<Header />);

let renderResult: RenderResult;

describe('Header:', () => {
  afterAll(cleanup);

  describe('Header with loading state:', () => {
    beforeEach(() => {
      if (renderResult?.unmount) {
        renderResult.unmount();
      }

      jest.mocked(useAutoConnectWallet).mockImplementation(() => 'idle');

      renderResult = renderHeaderWithProviders();
    });

    it('Show loader', () => {
      expect(getQueriesForElement(renderResult.getByTestId('header')).queryByTestId('loader')).toBeVisible();
    });
  });

  describe('Header with disconnected state:', () => {
    beforeEach(() => {
      if (renderResult?.unmount) {
        renderResult.unmount();
      }

      jest.mocked(useAutoConnectWallet).mockImplementation(() => 'attempted');

      renderResult = renderHeaderWithProviders();
    });

    it('Show loader', () => {
      expect(getQueriesForElement(renderResult.getByTestId('header')).queryByTestId('loader')).toBeVisible();
    });
  });

  describe('Header with connected state:', () => {
    beforeEach(() => {
      if (renderResult?.unmount) {
        renderResult.unmount();
      }

      jest.mocked(useAutoConnectWallet).mockImplementation(() => 'attempted');
      jest.mocked(useCurrentAccount).mockImplementation(createMockAccount);

      renderResult = renderHeaderWithProviders();
    });

    it('Don`t show loader', () => {
      expect(getQueriesForElement(renderResult.getByTestId('header')).queryByTestId('loader')).toBeNull();
    });

    it('Show account info', () => {
      expect(renderResult.queryByText(new RegExp(TEST_SHORT_WALLET_ACCOUNT_ADDRESS))).toBeVisible();
    });

    it('Show modal', () => {
      const logoutButton = getQueriesForElement(renderResult.getByTestId('header')).queryByRole('button');

      act(() => {
        logoutButton?.click();
      });

      expect(renderResult.queryByRole('dialog')).toHaveAttribute('open');
    });

    it('Snapshot', () => {
      expect(renderResult.container).toMatchSnapshot();
    });
  });
});
