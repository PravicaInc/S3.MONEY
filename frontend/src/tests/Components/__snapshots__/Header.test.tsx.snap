// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header: Header with connected state: Snapshot 1`] = `
<div>
  <style>
    [data-dapp-kit] {--dapp-kit-blurs-modalOverlay:blur(0);--dapp-kit-backgroundColors-primaryButton:#F6F7F9;--dapp-kit-backgroundColors-primaryButtonHover:#F0F2F5;--dapp-kit-backgroundColors-outlineButtonHover:#F4F4F5;--dapp-kit-backgroundColors-modalOverlay:rgba(24 36 53 / 20%);--dapp-kit-backgroundColors-modalPrimary:white;--dapp-kit-backgroundColors-modalSecondary:#F7F8F8;--dapp-kit-backgroundColors-iconButton:transparent;--dapp-kit-backgroundColors-iconButtonHover:#F0F1F2;--dapp-kit-backgroundColors-dropdownMenu:#FFFFFF;--dapp-kit-backgroundColors-dropdownMenuSeparator:#F3F6F8;--dapp-kit-backgroundColors-walletItemSelected:white;--dapp-kit-backgroundColors-walletItemHover:#3C424226;--dapp-kit-borderColors-outlineButton:#E4E4E7;--dapp-kit-colors-primaryButton:#373737;--dapp-kit-colors-outlineButton:#373737;--dapp-kit-colors-iconButton:#000000;--dapp-kit-colors-body:#182435;--dapp-kit-colors-bodyMuted:#767A81;--dapp-kit-colors-bodyDanger:#FF794B;--dapp-kit-radii-small:6px;--dapp-kit-radii-medium:8px;--dapp-kit-radii-large:12px;--dapp-kit-radii-xlarge:16px;--dapp-kit-shadows-primaryButton:0px 4px 12px rgba(0, 0, 0, 0.1);--dapp-kit-shadows-walletItemSelected:0px 2px 6px rgba(0, 0, 0, 0.05);--dapp-kit-fontWeights-normal:400;--dapp-kit-fontWeights-medium:500;--dapp-kit-fontWeights-bold:600;--dapp-kit-fontSizes-small:14px;--dapp-kit-fontSizes-medium:16px;--dapp-kit-fontSizes-large:18px;--dapp-kit-fontSizes-xlarge:20px;--dapp-kit-typography-fontFamily:ui-sans-serif, system-ui, -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, "Helvetica Neue", Arial, "Noto Sans", sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol", "Noto Color Emoji";--dapp-kit-typography-fontStyle:normal;--dapp-kit-typography-lineHeight:1.3;--dapp-kit-typography-letterSpacing:1;}
  </style>
  <div
    class="min-h-20 px-8 border-b border-borderPrimary bg-white flex items-center justify-center"
  >
    <div
      class="flex items-center justify-between max-w-screen-2xl mx-auto h-full w-full"
      data-testid="header"
    >
      <a
        href="/"
      >
        <svg />
      </a>
      <button
        class="border border-borderPrimary rounded-[10px] h-10 px-4 flex items-center justify-between gap-2 text-secondary font-medium transition hover:bg-actionPrimary hover:bg-opacity-10 hover:border-actionSecondary"
      >
        0x11...1111
        <svg />
      </button>
      <dialog
        class="backdrop:bg-modalBackdropColor rounded-xl p-6"
      >
        <div
          class="Toastify"
          id="modal-toast"
        />
        <div
          class=""
        >
          <p
            class="text-primary font-semibold text-2xl text-center mt-4"
          >
            Confirm Logout?
          </p>
          <div
            class="flex items-center justify-center gap-4 mt-8"
          >
            <button
              class="bg-white text-primary rounded-xl shadow-button border-borderPrimary border-solid border font-semibold flex items-center justify-center transition hover:border-actionSecondary hover:text-actionPrimary disabled:bg-slate-300 disabled:border-slate-400 disabled:text-white w-56 h-[56px]"
            >
              Cancel
            </button>
            <button
              class="text-white rounded-xl shadow-button border-actionSecondary border-solid border font-semibold flex items-center justify-center bg-buttonBgMain relative overflow-hidden after:content-[''] after:bg-buttonBgAfter after:w-full after:h-full after:absolute hover:bg-buttonBgAfter hover:text-actionPrimary transition disabled:bg-slate-300 disabled:bg-none disabled:border-slate-400 disabled:text-white w-56 h-[56px]"
            >
              Proceed
            </button>
          </div>
        </div>
      </dialog>
    </div>
  </div>
</div>
`;
